mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp

```diff
@@ -0,0 +1,491 @@
+         ++queueGroupIdx) {
+      const auto &group = queueGroupProperties[queueGroupIdx];
+      if (group.flags & ZE_COMMAND_QUEUE_GROUP_PROPERTY_FLAG_COMPUTE)
+        computeEngineOrdinal = queueGroupIdx;
+      else if (group.flags & ZE_COMMAND_QUEUE_GROUP_PROPERTY_FLAG_COPY) {
+        copyEngineOrdinal = queueGroupIdx;
+        copyEngineMaxMemoryFillPatternSize = group.maxMemoryFillPatternSize;
+      }
+      if (copyEngineOrdinal != -1u && computeEngineOrdinal != -1u)
+        break;
+    }
+    // Fallback to the default queue if no dedicated copy queue is available.
+    if (copyEngineOrdinal == -1u)
+      copyEngineOrdinal = computeEngineOrdinal;
+    assert(copyEngineOrdinal != -1u && computeEngineOrdinal != -1u &&
+           "Expected two engines to be available.");
+    ze_command_queue_desc_t cmdQueueDesc{
+        ZE_STRUCTURE_TYPE_COMMAND_QUEUE_DESC,
+        nullptr,
+        copyEngineOrdinal, // ordinal
+        0,                 // index (assume one physical engine in the group)
+        0,                 // flags
+        ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS,
+        ZE_COMMAND_QUEUE_PRIORITY_NORMAL};
+    L0_SAFE_CALL(zeCommandListCreateImmediate(context, device, &cmdQueueDesc,
+                                              &immCmdListCopy));
+    cmdQueueDesc.ordinal = computeEngineOrdinal;
+    L0_SAFE_CALL(zeCommandListCreateImmediate(context, device, &cmdQueueDesc,
+                                              &immCmdListCompute));
+  }
+  void cleanup() {
+    L0_SAFE_CALL(zeCommandListDestroy(immCmdListCopy));
+    L0_SAFE_CALL(zeCommandListDestroy(immCmdListCompute));
+    L0_SAFE_CALL(zeContextDestroy(context));
+  }
+  ~L0RtContext() { cleanup(); }
+};
+
+// L0 only supports pre-determined sizes of event pools,
+// implement a rt data struct to avoid running out of events.
+struct DynamicEventPool {
+  constexpr static size_t numEventsPerPool{128};
+  std::vector<ze_event_pool_handle_t> eventPools;
+  std::vector<ze_event_handle_t> availableEvents;
+  std::unordered_set<ze_event_handle_t> takenEvents;
+  size_t currentEventsLimit{0};
+  size_t currentEventsCnt{0};
+  L0RtContext *rtCtx;
+
+  DynamicEventPool(L0RtContext *rtCtx) : rtCtx(rtCtx) {
+    createNewPool(numEventsPerPool);
+  }
+
+  ~DynamicEventPool() {
+    assert(!takenEvents.size());
+    // zeEventDestroy will trigger L0_SAFE_CALL if an event is still used by
+    // device
+    for (auto event : availableEvents)
+      L0_SAFE_CALL(zeEventDestroy(event));
+    for (auto pool : eventPools)
+      L0_SAFE_CALL(zeEventPoolDestroy(pool));
+  }
+
+  void createNewPool(size_t numEvents) {
+    ze_event_pool_desc_t eventPoolDesc = {};
+    eventPoolDesc.flags = ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
+    eventPoolDesc.count = numEvents;
+    eventPools.push_back(nullptr);
+    L0_SAFE_CALL(zeEventPoolCreate(rtCtx->context, &eventPoolDesc, 1,
+                                   &rtCtx->device, &eventPools.back()));
+    currentEventsLimit += numEvents;
```

Comment:
What happens if the number of even and event pool surpass the "pre-determined sizes of event pools" supported by L0?

Reply:
I think a limit could be useful - just to be defensive in case something goes wrong.

Reply:
Added `maxEventsCount` which is set to 32K events. It should be enough.